crates/transpiler/src/passes/constrained_reschedule.rs

```diff
@@ -0,0 +1,304 @@
+///
+/// # Returns
+///
+/// A vector of `NodeIndex` values representing the immediate successor operation nodes.
+fn get_next_gate(dag: &DAGCircuit, node_index: NodeIndex) -> Vec<NodeIndex> {
+    dag.dag()
+        .neighbors_directed(node_index, Outgoing)
+        .filter(|&succ_idx| {
+            matches!(
+                dag.dag().node_weight(succ_idx),
+                Some(NodeType::Operation(_))
+            )
+        })
+        .collect()
+}
+
+/// Update the start time of the current node to satisfy alignment constraints.
+/// Immediate successors are pushed back to avoid overlap and will be processed later.
+///
+/// Note:
+/// This logic assumes that all bits in the qregs and cregs synchronously start and end,
+/// i.e. occupy the same time slot, but qregs and cregs can take different time slots
+/// due to classical I/O latencies.
+///
+/// # Args:
+/// * `py` - Python interpreter reference for PyO3 operations
+/// * `dag` - Reference to the DAGCircuit to be rescheduled with constraints
+/// * `node_index` - Index of the current node to be processed
+/// * `node_start_time` - Mutable Python dictionary mapping node indices to start times
+/// * `clbit_write_latency` - Additional latency for classical bit write operations
+/// * `pulse_align` - Alignment constraint for gate operations (in dt units)
+/// * `acquire_align` - Alignment constraint for measurement/reset operations (in dt units)
+/// * `target` - Optional target backend for duration information
+fn push_node_back(
+    dag: &DAGCircuit,
+    node_index: NodeIndex,
+    node_start_time: &Bound<PyDict>,
+    clbit_write_latency: u64,
+    pulse_align: u64,
+    acquire_align: u64,
+    target: Option<&Target>,
+) -> PyResult<()> {
+    let op = match dag.dag().node_weight(node_index) {
+        Some(NodeType::Operation(op)) => op,
+        _ => panic!("topological_op_nodes() should only return instances of DagOpNode."),
+    };
+
+    let op_view = op.op.view();
+    let alignment = match op_view {
+        OperationRef::Gate(_) | OperationRef::StandardGate(_) => Some(pulse_align),
+        OperationRef::StandardInstruction(StandardInstruction::Reset)
+        | OperationRef::StandardInstruction(StandardInstruction::Measure) => Some(acquire_align),
+        OperationRef::StandardInstruction(StandardInstruction::Delay(_)) => None,
+        _ => None,
+    };
+
+    let obj = node_start_time.get_item(node_index.index())?;
+    let mut this_t0: f64 = obj
+        .as_ref()
+        .ok_or_else(|| PyValueError::new_err("Missing value in node_start_time"))?
+        .extract()?;
+
+    if let Some(alignment) = alignment {
+        let misalignment = this_t0 % alignment as f64;
+        let shift = if misalignment != 0.0 {
+            (alignment as f64 - misalignment).max(0.0)
+        } else {
+            0.0
+        };
+        this_t0 += shift;
+        node_start_time
+            .set_item(node_index.index(), this_t0)
+            .unwrap();
+    }
```

Comment:
I'm surprised this is all floats, shouldn't this be an integer? I thought all the alignment constraints were in terms of `dt` which should all be in integers